src/rate_limiter.rs

```diff
@@ -173,97 +287,305 @@ impl RateLimiter {
+    }
+
+    // Return `true` if the outcome is as expected.
     fn check_message_addition(
         rate_limiter: &mut RateLimiter,
         client: &IpAddr,
-        large_msg: bool,
+        payload: &[u8],
         should_succeed: bool,
-    ) {
-        let payload: &[u8] = if large_msg {
-            &LARGE_MESSAGE
-        } else {
-            &SMALL_MESSAGE
-        };
+    ) -> bool {
         let hash = sha3_256(payload);
-        match rate_limiter.add_message(client, &hash, 2, 1, payload) {
+        let msg_id = MessageId::new();
+        match rate_limiter.add_message(client, &hash, &msg_id, 2, 1, payload) {
             Err(RoutingError::ExceedsRateLimit(returned_hash)) => {
                 if should_succeed {
-                    panic!("unexpected result");
+                    false
                 } else {
                     assert_eq!(hash, returned_hash);
+                    true
                 }
             }
             Ok(returned_len) => {
                 if should_succeed {
                     assert_eq!(payload.len() as u64, returned_len);
+                    true
                 } else {
-                    panic!("unexpected result");
+                    false
                 }
             }
-            _ => panic!("unexpected result"),
+            _ => false,
+        }
+    }
+
+    // Add a single `UserMessagePart` for a GET request to the rate limiter.
+    fn add_user_msg_part(
+        rate_limiter: &mut RateLimiter,
+        client: &IpAddr,
+        msg: &MessageContent,
+    ) -> Result<u64, RoutingError> {
+        if let MessageContent::UserMessagePart {
+            ref hash,
+            ref msg_id,
+            part_count,
+            part_index,
+            ref payload,
+            ..
+        } = *msg
+        {
+            rate_limiter.add_message(client, hash, msg_id, part_count, part_index, payload)
+        } else {
+            panic!("message is not a UserMessagePart: {:?}", msg);
+        }
+    }
+
+    // Send a single `UserMessagePart` for a response to the rate limiter for refunding.
+    fn refund_user_msg_part(
+        rate_limiter: &mut RateLimiter,
+        client: &IpAddr,
+        msg: &MessageContent,
+    ) -> Option<u64> {
+        if let MessageContent::UserMessagePart {
+            ref msg_id,
+            part_count,
+            part_index,
+            ref payload,
+            ..
+        } = *msg
+        {
+            rate_limiter.apply_refund_for_response(client, msg_id, part_count, part_index, payload)
+        } else {
+            panic!("message is not a UserMessagePart: {:?}", msg);
         }
     }
 
-    /// Checks that a single client cannot exceed its individual limit and that its throughput is
+    // Generate a single `UserMessagePart` for a random `GetIData` request.
+    fn random_payload() -> MessageContent {
+        let user_message = UserMessage::Request(Request::GetIData {
+            name: rand::random(),
```

Comment:
better not to use rand::random ? why not use a passed in SeededRng ?

also, try to use SeededRng in other places of rand::random as well?